### DISCUSSION TITLE
Using dbld to work on the syslog-ng source code

### DISCUSSION THREAD

**Discussion by amdrsantos** [syslog-ng/syslog-ng#4273] [Q&A] [answered]
I am trying to use the dockers to compile for Debian buster with the latest source code.

using the commands: 
> ./dbld/rules image-debian-buster
> ./dbld/rules shell-debian-buster

Inside the docker:
> ./autogen.sh
> mkdir -p build
> cd build
> ../configure

I hit:
> configure: WARNING: bison is found, but your bison version 3.3.2 is not recent enough, at least 3.7.6 is required
> configure: error: syslog-ng requires bison 3.7.6 or later. Your source tree seems to be from git, which doesn't have the bison generated files (like cfg-grammar.c). Please install/upgrade bison or use a distribu
> tion tarball.

I install the 3.7.6 and try to run configure again, but I get the following error:
> checking for suffix of object files... configure: error: in `/source/build/lib/ivykis':
> configure: error: cannot compute suffix of object files: cannot compile
> See `config.log' for more details
> configure: error: ../../../lib/ivykis/configure.gnu failed for lib/ivykis

I already have tried to do the same procedure with
> skip_submodules=1 ./autogen.sh

But I still hit:
> configure: error: Internal ivykis sources not found in lib/ivykis. This is a hard dependency, unable to build syslog-ng without them.

Is there any procedure that I can follow to compile using the dockers?

Thanks in advance

**Comment by bazsi** [comment, marked as answer]
The easiest is to use our "devshell" image, which has everything that you might need. Our non-devshell images are used to compile a release tarball, which may have some generated files already in them.

The problem above is that bison in Debian buster is relatively old and contains a bug that was fixed in 3.7.6 (and introduced IIRC in 3.7.1).

This is how I usually work with the syslog-ng source tree.

# Clone the github repo

```
$ git clone git@github.com:syslog-ng/syslog-ng.git
```

# Launch the devshell

You can run a shell inside the dev environment using "./dbld/rules shell", which starts a docker container, mounts the source/build/install directories inside the docker container and lets you hack on syslog-ng while being able to recompile/try it easily.

This is how this looks like for me:

```
$ ./dbld/rules shell
docker run -e USER_NAME_ON_HOST=bazsi --network=host --privileged --ulimit nofile=1024:1024 -v /home/bazsi/src/syslog-ng/syslog-ng:/source -v /home/bazsi/src/syslog-ng/syslog-ng/dbld:/dbld -v /home/bazsi/src/syslog-ng/syslog-ng/dbld/build:/build -v /home/bazsi/src/syslog-ng/syslog-ng/dbld/install:/install -e CONFIGURE_OPTS="${CONFIGURE_OPTS:---enable-debug --enable-manpages --with-python=3 --prefix=/install --disable-java --disable-manpages --enable-extra-warnings --enable-debug CFLAGS="-O0" --with-python=3}" -e CCACHE_DIR=/build/ccache -e MODE=snapshot -e VERSION=4.0.1.53.g1e97181 -e PATH=/usr/lib/ccache:/usr/lib64/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin -e GRADLE_USER_HOME=/build/gradle-home -e GRADLE_PROJECT_CACHE_DIR=/build/gradle-cache -e GRADLE_FLAGS=--build-cache -v /home/bazsi/.gitconfig:/build/.gitconfig -v /var/log/containers:/var/log/containers -v /var/log/pods:/var/log/pods -v /tmp/.X11-unix/:/tmp/.X11-unix -e DISPLAY=$DISPLAY -v ~/src/bpf:/reuse-bpf -v /var/log/journal:/var/log/journal -v /etc/machine-id:/etc/machine-id --rm -ti ghcr.io/syslog-ng/dbld-devshell /dbld/shell


Welcome to the syslog-ng developer's shell... (devshell)

You can find the source in /source, you can build the source using:

	$ cd /build
	$ /source/configure
	$ make
If configure is not yet available, run '/dbld/bootstrap' (in the container) or
'./dbld/rules bootstrap' (on the host) or manually execute autogen.sh in the 
source dir.

(dbld)bazsi@bzorp:/source$ 
```

My environment is not entirely pristine, our parameters to docker were customized to include extra settings. You can similarly customize  them using the ./dbld/rules.conf file, for me that file contains:

```
$ cat /dbld/rules.conf
DEFAULT_IMAGE=devshell
#CONFIGURE_ADD=--disable-java --disable-manpages --enable-extra-warnings --enable-debug --enable-pacct --enable-gcov CFLAGS="-O0"
CONFIGURE_ADD=--disable-java --disable-manpages --enable-extra-warnings --enable-debug CFLAGS="-O0" --with-python=3
MAKE_ARGS=-j8
#DOCKER_RUN_ARGS += -ti -v ~/src/syslog-ng-pe-project-7-0/zts:/zts -v ~/src/syslog-ng-pe-project-7-0/syslog-ng-pe-testdb:/testdb -v ~/src/syslog-ng-pe-project-7-0/syslog-ng-pe-modules:/pe-modules

DOCKER_RUN_ARGS += -v /var/log/containers:/var/log/containers -v /var/log/pods:/var/log/pods -v /tmp/.X11-unix/:/tmp/.X11-unix -e DISPLAY=$$DISPLAY -v ~/src/bpf:/reuse-bpf -v /var/log/journal:/var/log/journal -v /etc/machine-id:/etc/machine-id
```

rules.conf may override defaults in /dbld/rules, but its use is optional.

# Bootstrap the source

The source as stored in git does not contain some files that are automatically generated. autoconf, bison and other similar tools will generate some files that are used to build syslog-ng. To bootstrap using a just cloned source tree, you need to bootstrap it first.

Launch the devshell and once inside the embedded shell, invoke bootstrap:

```
$ ./dbld/rules shell
...
(dbld)bazsi@bzorp:~$ /dbld/bootstrap
...
```

The bootstrap takes a minute to run and will emit autoconf/automake/libtool etc output and run the configure scripts for you. syslog-ng source tree will be configured to use these directories within the docker image:
  * source is in /source (which is mounted from the outside syslog-ng source directory you just cloned)
  * build is in /build (which is mounted from syslog-ng/dbld/build)
  * install is in /install (which is mounted from syslog-ng/dbld/install)

Once bootstrap finishes, you can build and run syslog-ng inside the devshell while editing the source code outside.

# Build & install

```
$ cd /build
# you can run any kind of make targets, make install would copy binaries to /install
$ make -j install
```

# Run syslog-ng

If you installed syslog-ng in /install, you can run syslog-ng within the devshell, like this:

```
$ cd /install
# -F causes syslog-ng to run in the forefground
# -e to emit internal log messages to stderr
# -d to enable debug log level
# -f specifies the name of the configuration file
$ sbin/syslog-ng -Fed -f /install/etc/syslog-ng.conf
```

# Exiting the devshell

If you leave the bash shell inside the docker container, the container will be deleted, however all important state (/source, /build and /install directories) are mounted from the outside, so they will remain intact.

If you restart the container you will find the build products and installed files exactly where you left them.

# Additional actions in ./dbld/rules

The `./dbld/rules` script gives you a couple of actions, right within the source tree, without having to launch the devshell yourself.

```
$ ./dbld/rules help
This script allows you to build release/snapshot artifacts, such 
as rpm, deb packages and tarballs in controlled environments (e.g. 
within docker containers).

Please create the required docker image first using the 'image' target
then you can build syslog-ng binaries or do development using the 
targets described below.

Supported OSs are: centos-7 fedora-37 debian-stretch debian-buster debian-bullseye debian-testing debian-sid ubuntu-bionic ubuntu-focal ubuntu-jammy kira tarball devshell

Packaging targets:
  deb: generate debs in dbld/build for ubuntu-bionic
  rpm: generate RPMs in dbld/build for centos-7
  deb-<os>: generate debs in dbld/build for the specified OS
  rpm-<os>: generate rpms in dbld/build for the specified OS

Docker images:
  image: generate the default docker image
  image-<os>: generate docker image with a specific OS for building syslog-ng
  images: generate all docker images

Development:
  bootstrap: bootstrap source from a git clone (e.g. autogen.sh and friends)
  shell: start a shell in the specified OS
  login: start a 2nd shell alongside an already running one
  root-login: start a 2nd shell (as root) alongside an already running one
  tarball: generate a tarball in a controlled environment (e.g. docker)
  pkg-tarball: generate a tarball that includes deb/rpm packaging files
  prepare-release VERSION=x.y.z: prepare the source tree for release, does not commit
  release VERSION=x.y.z: generate a release tarball/deb package in a controlled environment (e.g. docker)

Notable Make variables:
  CONFIGURE_ADD -- add this to the configure command line in bootstrap
  DEFAULT_IMAGE -- override the default image, currently devshell
  TARBALL_IMAGE -- override the default image where the tarball is generated, currently tarball

```

Our GitHub actions based CI would use the `./dbld/rules` script to carry out various actions, such as generating the source tarball and building binary packages for various platforms.

**Reply by amdrsantos** [reply]
**Here are the steps that I used for compiling and packaging in debian-buster docker**

In one terminal, let's say Terminal 1:
> git clone https://github.com/syslog-ng/syslog-ng.git
> cd syslog-ng
>  
> #-# Check version
> scripts/version.sh
>  
> #-# Start the docker shell
> dbld/rules shell-debian-buster
>  
> ################# DOCKER #########################
> scripts/version.sh

In another terminal (Terminal 2): The bison utility needs to be updated to at least version 3.7.6. For that, a root console is needed in the docker.

> #-# Enter the docker in root mode
> dbld/rules root-login-debian-buster
>  
> ################# DOCKER #########################
> #-# Update the bison version
> wget --no-check-certificate http://ftp.gnu.org/gnu/bison/bison-3.7.6.tar.gz --output-document=/tmp/bison-3.7.6.tar.gz
> cd /tmp/
> tar -xvzf bison-3.7.6.tar.gz
> cd bison-3.7.6
> ./configure
> make install
>  
> #-# Clear the cache
> rm -rf /build/ccache

Continue in terminal (Terminal 1): Compile and package
> ################# DOCKER #########################
> /dbld/package
>  
> cd /dbld/build/debian-buster
> tar -czvf syslog-ng_4.0.1.64.gef55ed5_debian_buster.tar.gz *

**Reply by bazsi** [reply]
What you could do instead of deploying "bison" in Debian buster is to make you changes in devshell (which is Debian testing based), once you are happy, you can compile it on Debian buster (using ./dbld/rules deb-debian-buster).

In the background, dbld would generate the source tarball Debian testing and then use the resulting tarball within the Debian buster docker image.

This allows you to not deploy bison in the buster image, as the output of bison is included in the tarball.
